### ISSUE TITLE
FSE gutenberg_render_the_template breaks browser progressive rendering

### ISSUE THREAD

**Issue by tomjn** [WordPress/gutenberg#24378]
**Describe the bug**
The `gutenberg_render_the_template` function collects a pages content in a string then outputs it all at once. This means it suffers the same issues as Autoptimise and other plugins that use output buffers.

Previously this wasn't as big an issue as this only applied to the post content, but now that the entire page is output all at once, it's not possible for a browser to progressively render the page as it loads. All the browser can do is load the resources in the head and wait for the template content to finish generating.

This could be problematic, for example if a shortcode runs a query for 2 seconds, the entire page is held up an extra 2 seconds. Or if I make a remote request in a dynamic block at the end of the page, the browser gets nothing of the already generated output until everything is ready.

**To reproduce**
1. Create a site with FSE
2. Insert a dynamic block or shortcode that performs work that takes time
3. The browser will show nothing until it gets sent everything at once, degrading frontend performance and percieved loading times significantly

**Expected behavior**

Blocks would be output as they're generated/loaded so that output can be sent progressively, allowing the browser to do work while it's waiting for the rest of the content

**Comment by aristath** [comment]
What would it take for us to allow progressively rendering template-parts instead of combining them all under a `$_wp_current_template_content` global and render it? :thinking: 

I suppose the 1st step would be to introduce a new `gutenberg_render_the_template_part()` function, and then tweak the `gutenberg_render_the_template()` to get template-parts and call that new function for each one of them separately instead of running everything in a single pass?

**Comment by tomjn** [comment]
Possibly, but that only mitigates things in some situations.

The worst case scenario is multiple server side rendered blocks in a template that has no template parts, that would require displaying each block one at a time to resolve.

Currently, no output can be made until all blocks have been rendered to HTML anyway, the least we can do is change that to avoid a PHP global and render the blocks when they're output. Otherwise we can't even display the head and opening body tags with the current implementation until the very end

**Comment by aristath** [comment]
> Possibly, but that only mitigates things in some situations.
> 
> The worst case scenario is multiple server side rendered blocks in a template that has no template parts

Yeah, I agree... it would only make sense to do it if template-parts are used.

> that would require displaying each block one at a time to resolve.

OK... And would that cause any issues? I can't think of any. The content is its own separate block, so everything surrounding the content would be rendered one block at a time.
In fact, it could even boost performance a bit and if we add a few good filters it would also open the door to more opportunities. I can imagine for example someone would use an implementation like that + filters to render the content first and then bring in all other parts of the page asynchronously. I'm not saying that's necessarily a good thing, but it may be if used properly :thinking:

**Comment by tomjn** [comment]
We could parse the content then loop through the array of blocks
serialising and outputting them 1 by 1. It would take a little performance
hit but not much, the kind that doubles with every PHP upgrade.

At that point the worst case scenario is when nested blocks are used.

**Comment by talldan** [comment]
I think this should be a core trac issue now, as the functions are now part of WordPress core.

I see there's this one - https://core.trac.wordpress.org/ticket/43258. Maybe that's the same thing, and it was also recently fixed.
